Tests/TuistGeneratorTests/Graph/GraphCircularDetectorTests.swift

```diff
@@ -4,25 +4,190 @@ import XCTest
+        subject.start(from: a, to: c)
+        subject.start(from: c, to: d)
+        subject.start(from: d, to: e)
+        subject.start(from: d, to: b)
+        subject.start(from: e, to: a)
+
+        // Then
+        XCTAssertThrowsError(try subject.complete())
     }
 
-    func test_complete() throws {
-        try subject.start(from: node("a"), to: node("b"))
-        try subject.start(from: node("b"), to: node("c"))
-        try subject.start(from: node("b"), to: node("d"))
-        subject.complete(node("a"))
+    func test_noCycleDetected_1() throws {
+        // Given
+        let a = node("a")
+        let b = node("b")
+        let c = node("c")
+        let d = node("d")
+
+        // When
+        subject.start(from: a, to: b)
+        subject.start(from: b, to: c)
+        subject.start(from: b, to: d)
+
+        // Then
+        XCTAssertNoThrow(try subject.complete())
+    }
+
+    func test_noCycleDetected_2() throws {
+        // Given
+        let a = node("a")
+        let b = node("b")
+        let c = node("c")
+
+        // When
+        subject.start(from: a, to: b)
+        subject.start(from: c, to: b)
+        subject.start(from: c, to: a)
+
+        // Then
+        XCTAssertNoThrow(try subject.complete())
+    }
+
+    func test_noCycleDetected_3() throws {
+        // Given
+        let a = node("a")
+        let b = node("b")
+        let c = node("c")
+        let d = node("d")
+        let e = node("e")
+
+        // When
+        subject.start(from: a, to: b)
+        subject.start(from: a, to: c)
+        subject.start(from: a, to: d)
+        subject.start(from: a, to: e)
+
+        subject.start(from: b, to: c)
+        subject.start(from: b, to: d)
+        subject.start(from: b, to: e)
+
+        subject.start(from: c, to: d)
+        subject.start(from: c, to: e)
+
+        subject.start(from: d, to: e)
+
+        // Then
+        XCTAssertNoThrow(try subject.complete())
+    }
+
+    func test_noCycleDetected_detachedGraphs() throws {
+        // Given
+        let a = node("a")
+        let b = node("b")
+        let c = node("c")
+        let d = node("d")
+        let e = node("e")
+
+        // When
+        subject.start(from: a, to: b)
+        subject.start(from: a, to: c)
+        subject.start(from: b, to: c)
+
+        subject.start(from: d, to: e)
+
+        // Then
+        XCTAssertNoThrow(try subject.complete())
+    }
+
+    func test_cycleDetected_detachedGraphs() throws {
+        // Given
+        let a = node("a")
+        let b = node("b")
+        let c = node("c")
+        let d = node("d")
+        let e = node("e")
+        let f = node("f")
+
+        // When
+        subject.start(from: a, to: b)
+        subject.start(from: a, to: c)
+        subject.start(from: b, to: c)
+
+        subject.start(from: d, to: e)
+        subject.start(from: e, to: f)
+        subject.start(from: f, to: d)
+
+        // Then
+        XCTAssertThrowsError(try subject.complete())
```

Comment:
I'd assert that we are throwing the right error.

Reply:
👍 indeed otherwise it could be flagging the wrong nodes! will update it